nasa/earthdata-varinfo
# MSD-4058 - Replace native ID  spaces with underscores.

## Description

This PR supports ASDC, who are trying to use `earthdata-varinfo` to generate UMM-Var records for variables that include spaces in their full paths. This is currently causing issues in the choice of CMR native ID. To support this, spaces in generated native IDs are converted to underscores (in a similar manner to how slashes are already converted to underscores).

As a bonus, this PR relaxes the `numpy` dependency, which theoretically allows for support under Python 3.12.

~~This cannot be rigorously tested via our CI/CD at this time, because one of the dependencies our testing uses (`xmlrunner`) does not work under Python 3.12. However, continued operations under Python 3.9, 3.10 and 3.11 can be confirmed.~~ 

**Update:** I found you can run a test suite that utilises `unittest` style classes with `pytest`, and so use the plugins from the `pytest` ecosystem. I've swapped over to running our `unittest` style test suite with `pytest`, which allows us to use the `--junitxml` option. This does not use `unittest-xml-runner` and is compatible with Python 3.12 :tada:. This PR also now uses `pytest-cov` to generate coverage reports as before.

Generally, providing a broader range of options for `numpy` enables `earthdata-varinfo` to be compatible with more client Python environments. The relaxed constraint continues to support the previously pinned version at a minimum.

## Jira Issue ID

MSD-4058 (not part of the regular ESDIS/EED Jira instance)

## Local Test Steps

* Pull this branch.
* Use `earthdata-varinfo` as you would have before, to make sure there are no regressions.
* Try generating some UMM-Var records for a file with spaces in the variables names. Then use the `generate_variable_native_id` function to confirm that any spaces are converted to underscores.

## PR Acceptance Checklist
* ~~Jira ticket acceptance criteria met.~~
* ~~Documentation updated (if needed).~~
